### DISCUSSION TITLE
ESP32 BOX Lite - Writing to the Screen

### DISCUSSION THREAD

**Discussion by jouellnyc** [micropython/micropython#10435] [Hardware & Peripherals] [answered]
Good day MicroPython'ers.
 
Just starting out with an [ESP 32 Box lite ](https://www.espressif.com/en/news/ESP32-S3-BOX_video) and loaded MP with it with some help from this forum a few days ago.

I've got a repl and can connect to the internet which is great, but trying to print to the screen now. 

I thought the screen was an ili9341 -- I don't really know -- as I have not found that information online explicitly, but  grepping the repo https://github.com/espressif/esp-box.git for clues I found something:

```
$ grep -ir ili934   .
./docs/_static/_table/specs_for_box.md:    <td class="tg-ycr8">ILI9342C</td>
```

I tried using this excellent driver: https://github.com/rdagger/micropython-ili9341 which works great for my other esp32s connected to ili9341's.

I'm creating the spi object like this:

`spi = SPI(1, baudrate=40000000, sck=Pin(12), mosi=Pin(11) , miso=Pin(13))
`

as that is what the back of the device says the values are. And then a display object:

`display = Display(spi, dc=Pin(4), cs=Pin(5), rst=Pin(48), width=width, height=height, rotation=rotation)`

based on these values:

GPIO_LCD_CS =     (GPIO_NUM_5),
GPIO_LCD_RST =    (GPIO_NUM_48),
GPIO_LCD_DC =     (GPIO_NUM_4),

from here:

https://github.com/espressif/esp-box/blob/master/components/bsp/src/boards/esp32_s3_box_lite.c

I've tried lots of different combinations for the spi, and google/chatgpt'ed based on some educated guess but no real  luck just yet. 

Not sure how popular this board is, but anyone know if I am way off or if I completely missed something?

EDIT:(just saw this now https://github.com/russhughes/ili9342c_mpy, seems like a full firmware and not specific to the S3 Box Lite, but might be worth checking out as it seems to be for "ILI9342C" vs "ILI9341")

**Comment by russhughes** [comment, marked as answer]
I don't have one of these yet, so I haven't tested any of this. According to the Hardware Overview the S3 Box Lite has an ESP32-S3 with a st7789 display. You could try the (untested) ESP32_BOX_LITE [firmware](https://github.com/russhughes/st7789_mpy/blob/master/firmware/ESP32_BOX_LITE/firmware.bin) from https://github.com/russhughes/st7789_mpy.  Configuration Example (untested) [tft_config.py](https://github.com/russhughes/st7789_mpy/blob/master/examples/configs/esp32_box_lite/tft_config.py).

Display I/O Pins: 4 DC, 5 CS, 6 MOSI, 7 CLK, 48 RESET and 45 BACKLIGHT

References:
[Hardware Overview](https://github.com/espressif/esp-box/blob/master/docs/hardware_overview/esp32_s3_box_lite/hardware_overview_for_lite.md)
[Schematic](https://github.com/espressif/esp-box/blob/master/hardware/esp32_s3_box_lite_MB_V1.1/schematic/SCH_ESP32-S3-BOX-Lite_MB_V1.1_20211221.pdf)

**Comment by jouellnyc** [comment]
Thanks very much. Here's what I found in case others had my experience:

- You're right on the Display Pins Numbers - The numbers on the back of the Box itself must be for the available / exposed SPI bus, where the LCD is unavailable / not exposed.

- I used those numbers for the SPI bus line you shared and that were in the doc: 
![image](https://user-images.githubusercontent.com/32470508/211219158-5c757807-9790-4304-927a-8a02c1d4448e.png)
and I was able to write to the screen with the ili9341 driver. 

However,  trying to fill the screen w/"green" comes up with more of a "pink" color (using the color565()  method in the ili9341.py module of the same driver). Add oddly enough white displays black and vice versa. I am guessing it's because that driver wasn't explicitly written for that piece of hardware, or maybe I need to research it more from the rgb standpoint --  I was only able to  spend an hour+ or so on it.

- Thank you very much for the firmware. I did try it out and for some reason it didn't work boot. It gave "invalid header" repeatedly.  I wasn't able to capture more than that, when I use gtkterm it was scrolling too fast. I did try erasing the flash and different modes like dio/dout/qio. But it's likely something that I have done versus your firmware.

I plan on more research. Will reply back w/any more success. Thanks again very much for the help!

**Reply by jouellnyc** [reply]
Update: basically I can do everything I want to achieve using the ili934 driver given the above small caveats, so no real update/change to the post. I didn't actually see an explicit  st7789.py file, so I guess my question is (you'll have to forgive the naive question): Is ridiculous to think there is a st7789.py file? I'm guessing that's 'baked' in the the frozen firmware.

**Reply by russhughes** [reply]
Yes, the driver is written in C and compiled into the Firmware. See 
https://github.com/russhughes/st7789py_mpy for a pure python driver with less functionality that could easily be modified.

**Reply by jouellnyc** [reply]
OK. Thank you for that! Much appreciated! The reason I ask is the screen is 'ghosting' where text does not get 'overwritten' properly. Since  I am not using the correct driver, I don't really know how to think about it. I'll certainly given the file above a try.

I was considering  the ESP32 S3 Box which is using a ili9342c display - I'm guessing [this](https://github.com/russhughes/ili9342c_mpy/tree/main/firmware) could work just fine for that?
But also guessing it's likely untested like the S3 box lite. (I have a second box lite I could send to you if that helps with supporting the box lite)

Given I spent so much time on the S3  box lite I was hoping to finish my project on it, given the time it would take to recode, but let see..

**Reply by russhughes** [reply]
I now have an ESP32 S3 Box, which works with the driver.  The display orientation is rotated 180' from what my driver typically expects, so the scrolling demos don't look right unless you configure the display upside down.  See https://github.com/russhughes/st7789_mpy/blob/4d7f2b60d60b739325502285987c3830814d7688/examples/configs/esp32_box_lite/tft_config.py for an example tft_config.py that works with the display. This should work with both the lite and non-lite versions.

**Comment by jouellnyc** [comment]
Oh great to hear you have the ESP32 S3 Box. Sorry for my ignorance here but what firmware are you using for it?

For the Box Lite, I tried the tft_confg.py and used the  [less functional driver](https://github.com/russhughes/st7789py_mpy/blob/master/lib/st7789py.py).

I got some errors - it seems like it does not support "custom_init=custom_init"  in the ST7789 class and the variable 'BGR' went missing. But I think that's what you mean but it was less functional. I tried to remove some of the init params , did not quite get it to work yet, but trying ...

**Reply by russhughes** [reply]
I use the firmware from: https://github.com/russhughes/st7789_mpy/blob/4d7f2b60d60b739325502285987c3830814d7688/firmware/ESP32_BOX_LITE/firmware.bin

The python driver does not have all the features included in the C version, custom_init being one of those.

**Reply by jouellnyc** [reply]
OK thanks for confirming!

**Reply by jouellnyc** [reply]
I got an S3 Box (non-lite) and it's quite a nice demonstration you created with the examples like 'hello.py'! 
I can write to the screen fine but am trying to add a helper method to it the instance of 'tft', using tft_config.py:


```
import st7789
import tft_config
tft = tft_config.config(0)

def helper():
    tft.text()
    tft.fill_circle()

tft.helper = helper     #I get a  "has no attribute" error here

``` 

Is the st7789.ST7789  class 'frozen'?

**Comment by russhughes** [comment]
Yes, it is compiled 'C' code.

**Reply by jouellnyc** [reply]
OK. Thanks. In my mind I thought I could modify the instance of the class in RAM. Thanks for the quick reply!
